### PR TITLE
[APOLLO-2103] Fix SSRF

### LINKED ISSUE #2103: 接口"/system-info/health"存在SSRF漏洞
# 问题说明
接口“/system-info/health”源码使用RestTemplate发起请求，对用户输入的host没有限制。
源码
```java
@PreAuthorize(value = "@permissionValidator.isSuperAdmin()")
@GetMapping(value = "/health")
public Health checkHealth(@RequestParam String host) {
       return restTemplate.getForObject(host + "/health", Health.class);
}
```

# 利用
## 1. 对内网发起请求，打poc
对内网地址可以搭poc，在host末尾通过添加“%23“截断源码中的“/health”
证明：本地搭建一个有命令执行的php服务，和一个apollo服务，通过接口向php服务发起请求，命令执行成功。

## 2. 内网端口探测
参数host传入内网ip和端口，可通过返回的错误信息判断端口是否开启。
证明：官方给的测试环境106.12.25.204:8070, 利用接口访问局域网内192.168.0.4:8080，返回证明可以从外网访问内网ip。

### PR DESCRIPTION
**Pull request by kezhenxu94**
fix SSRF, resolve #2103

### CODE REVIEW

File: apollo-portal/src/main/resources/static/scripts/controller/SystemInfoController.js [hunk 1]
```diff
@@ -28,13 +28,13 @@ function SystemInfoController($scope, toastr, AppUtil, AppService, ClusterServic
...
+    function check(instanceId) {
+        SystemInfoService.check_health(instanceId).then(function (result) {
             var status = result.status.code;
-            if (status == 'UP') {
-                toastr.success(host + ' is healthy!');
+            if (status === 'UP') {
+                toastr.success(instanceId + ' is healthy!');
             } else {
-                toastr.error(host + ' is not healthy, please check ' + host + '/health for more information!');
+                toastr.error(instanceId + ' is not healthy, please check /health of ' + instanceId + 'for more information!');
```
Comment (nobodyiam) [inline apollo-portal/src/main/resources/static/scripts/controller/SystemInfoController.js]: I think the host is more user friendly in the error message. How about we pass both host and instanceId into this function and use instanceId to request the backend while using host in the toaster?

File: apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/controller/SystemInfoController.java [hunk 2]
```diff
@@ -84,8 +68,52 @@ public SystemInfo getSystemInfo() {
...
+
+    String selectedMetaServerAddress = MetaDomainConsts.getDomain(env);
+    try {
+      environmentInfo.setConfigServices(getServerAddress(selectedMetaServerAddress, CONFIG_SERVICE_URL_PATH));
+
+      environmentInfo.setAdminServices(getServerAddress(selectedMetaServerAddress, ADMIN_SERVICE_URL_PATH));
+    } catch (Throwable ex) {
+      String errorMessage = "Loading config/admin services from meta server: " + selectedMetaServerAddress + " failed!";
+      logger.error(errorMessage, ex);
+      environmentInfo.setErrorMessage(errorMessage + " Exception: " + ex.getMessage());
```
Comment (nobodyiam) [inline apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/controller/SystemInfoController.java]: In this case, environmentInfo.getAdminServices and environmentInfo.getConfigServices will return null and there will be null pointer exception in line 77 and line 83

**Comment by nobodyiam** [comment]
@liutao5121 

What do you mean by `/health 未授权访问漏洞 解决了吗？`?
I believe `/health` doesn't expose any sensitive information, and it could be used by external platforms like Kubernetes readiness probe.
But if you really want to close this endpoint, you may simply configure it via the standard [spring boot way](https://docs.spring.io/spring-boot/docs/current/reference/html/actuator.html#actuator.endpoints.enabling).
Currently, apollo [exposes](https://github.com/apolloconfig/apollo/blob/master/apollo-common/src/main/resources/application.properties#L37) `info,health,metrics,prometheus` by default

**Comment by liutao5121** [comment]
> 

I tried to resolve the unauthorized access vulnerability (/health) by turning off the health endpoint of the ConfigService module, which caused no configuration information to be seen on the Apollo-Portal interface

**Comment by liutao5121** [comment]
image

**Comment by liutao5121** [comment]
If the configuration "management. Endpoints. Web. Exposure. Exclude = health", the upper right corner of the portal page complains, don't see any configuration information

**Comment by nobodyiam** [comment]
Right...I forgot the portal will ping admin-service's health endpoint to check whether it's healthy.

https://github.com/apolloconfig/apollo/blob/d15331b25242c5a5d80118b139e7a27bb95076bb/apollo-portal/src/main/java/com/ctrip/framework/apollo/portal/api/AdminServiceAPI.java#L42-L44

**Pull request merged**